### PR TITLE
add stream mode which do not up dfget progress.

### PR DESCRIPTION
**Pull request by wangforthinker**
Signed-off-by: allen.wq <allen.wq@alipay.com>

### Ⅰ. Describe what this PR did
In the past, if request the file, dfdaemon should boot the dfget as a progress to register and pull file from peer node. To speed up the p2p download rate, we extract dfget lib as the inner progress in dfdaemon progress.

### Ⅱ. Does this pull request fix one issue?

NONE.

### Ⅲ. Why don't you add test cases (unit test/integration test)? (你真的觉得不需要加测试吗？)

Add test.

### CODE REVIEW

File: dfdaemon/downloader/downloader.go [hunk 1]
```diff
@@ -25,5 +28,12 @@ type Interface interface {
 	DownloadContext(ctx context.Context, url string, header map[string][]string, name string) (string, error)
 }
 
+type Stream interface {
+	// DownloadContext downloads the resource as specified in url, and it accepts
+	// a context parameter so that it can handle timeouts correctly.
+	DownloadStreamContext(ctx context.Context, url string, header map[string][]string, name string) (io.Reader, error)
```
Comment (wangforthinker) [inline dfdaemon/downloader/downloader.go]: name字段删掉。

File: dfdaemon/transport/transport.go [hunk 2]
```diff
@@ -161,6 +183,21 @@ func (roundTripper *DFRoundTripper) downloadByGetter(ctx context.Context, url st
...
+func (roundTripper *DFRoundTripper) downloadByStream(ctx context.Context, url string, header map[string][]string, name string) (*http.Response, error) {
+	logrus.Infof("start download url:%s to %s in repo", url, name)
+	reader, err := roundTripper.StreamDownloader.DownloadStreamContext(ctx, url, header, name)
+	if err != nil {
+		logrus.Errorf("download fail: %v", err)
+		return nil, err
+	}
+
+	resp := &http.Response{
+		StatusCode: 200,
```
Comment (wangforthinker) [inline dfdaemon/transport/transport.go]: status code考虑设置为200或206.

Comment (wangforthinker) [reply dfdaemon/transport/transport.go]: 回源获取到的resp header，可以保存下来，统一返回给request。

**Comment by hhhhsdxxxx** [comment]
lgtm!

File: cmd/dfdaemon/app/init.go [hunk 3]
```diff
@@ -17,34 +17,67 @@
...
+	"github.com/dragonflyoss/Dragonfly/pkg/netutils"
 	statutil "github.com/dragonflyoss/Dragonfly/pkg/stat"
+	"github.com/dragonflyoss/Dragonfly/pkg/stringutils"
 
 	"github.com/pkg/errors"
 	"github.com/sirupsen/logrus"
 )
 
+// adjustSupernodeList adjusts the super nodes [a,b] to [a,b,b,a]
+func adjustSupernodeList(nodes []string) []string {
```
Comment (xujihui1985) [inline cmd/dfdaemon/app/init.go]: adjust SupernodeList的目的？

File: cmd/dfdaemon/app/init.go [hunk 4]
```diff
@@ -17,34 +17,67 @@
...
+	default:
+		util.Shuffle(nodesLen, func(i, j int) {
+			nodes[i], nodes[j] = nodes[j], nodes[i]
+		})
+		return append(nodes, nodes...)
+	}
+}
+
+// checkConnectSupernode return the localIP which connects to supper node
+func checkConnectSupernode(nodes []string) (localIP string) {
```
Comment (xujihui1985) [inline cmd/dfdaemon/app/init.go]: 是否调整一下函数名，比如getLocalIP

File: cmd/dfdaemon/app/init.go [hunk 5]
```diff
@@ -58,15 +91,13 @@ func initDfdaemon(cfg config.Properties) error {
...
 		)
 	}
+	cfg.SuperNodes = adjustSupernodeList(cfg.SuperNodes)
+	if stringutils.IsEmptyStr(cfg.LocalIP) {
+		cfg.LocalIP = checkConnectSupernode(cfg.SuperNodes)
+	}
 
 	go cleanLocalRepo(cfg.DFRepo)
 
-	dfgetVersion, err := exec.Command(cfg.DFPath, "version").CombinedOutput()
```
Comment (xujihui1985) [inline cmd/dfdaemon/app/init.go]: 非streammode下，是否还需要保留原来逻辑

File: cmd/dfdaemon/app/root.go [hunk 6]
```diff
@@ -72,6 +72,10 @@ var rootCmd = &cobra.Command{
 		if err != nil {
 			return errors.Wrap(err, "create dfdaemon from config")
 		}
+		// if stream mode, launch peer server in dfdaemon progress
+		if cfg.StreamMode {
+			go dfdaemon.LaunchPeerServer(*cfg)
```
Comment (xujihui1985) [inline cmd/dfdaemon/app/root.go]: 这里是否要加个退出signal

Comment (wangforthinker) [reply cmd/dfdaemon/app/root.go]: 已经有信号退出机制了。

File: cmd/dfdaemon/app/init.go (hunk 4 above)
Comment (wangforthinker) [reply cmd/dfdaemon/app/init.go]: done.

File: cmd/dfdaemon/app/init.go (hunk 5 above)
Comment (wangforthinker) [reply cmd/dfdaemon/app/init.go]: done

**Comment by lowzj** [comment]
please resolve the `linters-check` problemt

**Comment by wangforthinker** [comment]
> please resolve the `linters-check` problemt

FILE: ./ROADMAP.md
[✖] https://coreos.com/operators/ → Status: 0 Error: write EPROTO 140270879573824:error:1407742E:SSL routines:SSL23_GET_SERVER_HELLO:tlsv1 alert protocol version:../deps/openssl/openssl/ssl/s23_clnt.c:802:


9 links checked.

ERROR: 1 dead links found!

It looks like in ROADMAP.md, it has dead links.

**Pull request merged**